src/silo/query_engine/actions/aa_mutations.cpp

```diff
@@ -1,25 +1,195 @@
 #include "silo/query_engine/actions/aa_mutations.h"
 
+#include <cmath>
+#include <map>
+#include <utility>
+#include <vector>
+
+#include <tbb/blocked_range.h>
+#include <tbb/parallel_for.h>
 #include <nlohmann/json.hpp>
 
+#include "silo/common/aa_symbols.h"
+#include "silo/database.h"
 #include "silo/query_engine/operator_result.h"
 #include "silo/query_engine/query_parse_exception.h"
 #include "silo/query_engine/query_result.h"
+#include "silo/storage/aa_store.h"
+#include "silo/storage/database_partition.h"
+
+using silo::query_engine::OperatorResult;
+
+namespace {
+
+std::pair<std::vector<size_t>, std::vector<size_t>> preFilterBitmaps(
+   const silo::AAStore& aa_store,
+   std::vector<OperatorResult>& bitmap_filter
+) {
+   std::vector<size_t> bitmap_filters_to_evaluate;
+   std::vector<size_t> full_bitmap_filters_to_evaluate;
+   for (size_t i = 0; i < aa_store.partitions.size(); ++i) {
+      const silo::AAStorePartition& aa_store_partition = aa_store.partitions.at(i);
+      OperatorResult& filter = bitmap_filter[i];
+      const size_t cardinality = filter->cardinality();
+      if (cardinality == 0) {
+         continue;
+      }
+      if (cardinality == aa_store_partition.sequence_count) {
+         full_bitmap_filters_to_evaluate.push_back(i);
+      } else {
+         if (filter.isMutable()) {
+            filter->runOptimize();
+         }
+         bitmap_filters_to_evaluate.push_back(i);
+      }
+   }
+   return {bitmap_filters_to_evaluate, full_bitmap_filters_to_evaluate};
+}
+
+}  // namespace
 
 namespace silo::query_engine::actions {
 
-AAMutations::AAMutations() = default;
+AAMutations::AAMutations(std::string aa_sequence_name, double min_proportion)
+    : aa_sequence_name(std::move(aa_sequence_name)),
+      min_proportion(min_proportion) {}
+
+std::array<std::vector<uint32_t>, AAMutations::MUTATION_SYMBOL_COUNT> AAMutations::
+   calculateMutationsPerPosition(
+      const AAStore& aa_store,
+      std::vector<OperatorResult>& bitmap_filter
+   ) {
+   const size_t sequence_length = aa_store.reference_sequence.length();
+
+   std::vector<size_t> bitmap_filters_to_evaluate;
+   std::vector<size_t> full_bitmap_filters_to_evaluate;
+   std::tie(bitmap_filters_to_evaluate, full_bitmap_filters_to_evaluate) =
+      preFilterBitmaps(aa_store, bitmap_filter);
+
+   std::array<std::vector<uint32_t>, MUTATION_SYMBOL_COUNT> count_of_mutations_per_position;
+   for (auto& vec : count_of_mutations_per_position) {
+      vec.resize(sequence_length);
+   }
+   static constexpr int POSITIONS_PER_PROCESS = 300;
+   const tbb::blocked_range<uint32_t> range(
+      0, sequence_length, /*grain_size=*/POSITIONS_PER_PROCESS
+   );
+   tbb::parallel_for(range.begin(), range.end(), [&](uint32_t pos) {
+      for (const size_t partition_index : bitmap_filters_to_evaluate) {
```

Comment:
Can we put this for loop and the next two into a function with a meaningful name? It seems this would make the comments redundant and the code more readable

Reply:
At first I though it would be too hard to find a nice abstraction, but I think I found a nice refactor